src/hooks.ts

```diff
@@ -36,3 +42,111 @@ export const useHasResourcePermission = (resource: Resource, authzUrl: string, p
     const { permissions, isFetching } = useResourcePermissions(resource, authzUrl) ?? {};
     return { isFetching, hasPermission: permissions.includes(permission) };
 };
+
+export const useOpenIdConfig = (openIdConfigUrl: string) => {
+    const dispatch = useDispatch();
+
+    useEffect(() => {
+        dispatch(fetchOpenIdConfiguration(openIdConfigUrl));
+    }, [dispatch, openIdConfigUrl]);
+
+    return useSelector((state: RootState) => state.openIdConfiguration.data);
+}
+
+export const useSignInPopupTokenHandoff = (
+    applicationUrl: string,
+    authCallBackUrl: string,
+    clientId: string,
+    windowMessageHandler: MutableRefObject<null | MessageHandlerFunc>
+) => {
+    const dispatch = useDispatch();
+    useEffect(() => {
+        if (windowMessageHandler.current) {
+            window.removeEventListener("message", windowMessageHandler.current);
+        }
+        windowMessageHandler.current = (e: MessageEvent) => {
+            if (e.origin !== applicationUrl) return;
+            if (e.data?.type !== AUTH_RESULT_TYPE) return;
+            const { code, verifier } = e.data ?? {};
+            if (!code || !verifier) return;
+            localStorage.removeItem(LS_SIGN_IN_POPUP);
+            dispatch(tokenHandoff({ code, verifier, clientId: clientId, authCallbackUrl: authCallBackUrl }));
+        };
+        window.addEventListener("message", windowMessageHandler.current);
+    }, [dispatch]);
+};
+
+export const useSessionWorkerTokenRefresh = (
+    clientId: string,
+    sessionWorkerRef: MutableRefObject<null | Worker>,
+    createWorker: () => Worker,
+    fetchUserDependentData: (servicesCb: () => void) => void,
+) => {
+    const dispatch = useDispatch();
+    useEffect(() => {
+        if (!sessionWorkerRef.current) {
+            const sw = createWorker();
+            sw.addEventListener("message", () => {
+                dispatch(refreshTokens(clientId));
+                dispatch(fetchUserDependentData);
+            });
+            sessionWorkerRef.current = sw;
+        }
+
+        return () => {
+            if (sessionWorkerRef.current) {
+                sessionWorkerRef.current.terminate();
+            }
+        }
+    }, [dispatch, createWorker, fetchUserDependentData]);
+};
+
+interface OpenIdConfig {
+    authorization_endpoint: string;
+    [x: string | number | symbol]: unknown;
+}
+
+export const useOpenSignInWindowCallback = (
+    signInWindow: MutableRefObject<null | Window>,
+    openIdConfig: OpenIdConfig,
+    clientId: string,
+    authCallbackUrl: string,
+    windowFeatures = "scrollbars=no, toolbar=no, menubar=no, width=800, height=600"
+) => {
+    return useCallback(() => {
+        if (signInWindow.current && !signInWindow.current.closed) {
+            signInWindow.current.focus();
+            return;
+        }
+    
+        if (!openIdConfig || !window.top) return;
```

Comment:
actually, shouldn't this hook `useOpenIdConfig` instead? (outside the callback, instead of accepting it as a parameter)